File: leetcode/126.cpp
class Solution {
public:
	vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
		vector<vector<string>> ret;
		map<string, vector<string>> graph;
		map<string, int> visited;
		auto index = std::find(wordList.begin(), wordList.end(), beginWord);
		if (index == wordList.end()) wordList.push_back(beginWord);
		construct_graph(wordList, graph);
		bfs(endWord, beginWord, graph, visited, ret);
		return ret;
	}

	int bfs(string &beginWord, string &endWord, map<string, vector<string>>& graph, map<string, int> &visited, vector<vector<string>> &ret) {
		vector<string> item;
		vector<tuple<string, int, int>> q;
		q.push_back(tuple(beginWord, 1, -1));
		visited.insert(pair(beginWord, 1));
		tuple<string, int, int> cur;

		int minLen = INT_MAX;
		bool isfind = false;

		int j = 0;

		string key;
		int step;
		int pos;

		while (j < q.size()) {
			cur = q[j];

			key = get<0>(cur);
			step = get<1>(cur);
			pos = get<2>(cur);

			if (step > minLen) break;

			if (key == endWord) {
				if (false == isfind) {
					minLen = get<1>(cur);
					isfind = true;
				}
				item.clear();
				item.push_back(key);
				tuple<string, int, int> pre;
				while (pos != -1) {
					pre = q[pos];
					item.push_back(get<0>(pre));
					pos = get<2>(pre);
				}
				//    reverse(item.begin(), item.end());
				ret.push_back(item);
			}

			for (int i = 0; i < graph[key].size(); i++) {
				string nextKey = graph[key][i];
				if (visited.find(graph[key][i]) == visited.end() || visited[nextKey] == step + 1) {
					q.push_back(tuple(nextKey, step + 1, j));
					visited.insert(pair(nextKey, step + 1));
				}
			}
			j++;
		}
		return 0;
	}

	bool connect(string &word1, string &word2) {
		int count = 0;
		for (int i = 0; i < word1.size(); i++) {
			if (word1[i] != word2[i]) count++;
		}
		return count == 1;
	}

	void construct_graph(vector<string>& wordList, map<string, vector<string>>& graph) {
		for (int i = 0; i < wordList.size() - 1; i++) {
			for (int j = i + 1; j < wordList.size(); j++) {
				if (connect(wordList[i], wordList[j])) {
					graph[wordList[i]].push_back(wordList[j]);
					graph[wordList[j]].push_back(wordList[i]);
				}
			}
		}

	}
};

File: leetcode/126.java

class Solution {
    class Pair{
        String first;
        int step;
        int index;
        Pair(String first, int step, int index){
            this.first = first;
            this.step = step;
            this.index = index;
        }
    }

    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        List<List<String>> ret = new ArrayList<>();
        Map<String, List<String>> graph = new HashMap<>();
        Map<String, Integer> visited = new HashMap<>();

        if(!wordList.contains(beginWord)) wordList.add(beginWord);

        for(int i=0; i<wordList.size(); i++) {
            graph.put(wordList.get(i), new ArrayList<String>());
        }
        construct_graph(wordList, graph);
        bfs( endWord, beginWord, graph, visited, ret);
        return ret;
    }

    void bfs(String beginWord, String endWord, Map<String, List<String>> graph, Map<String, Integer> visited, List<List<String>> ret ){
        List<String> list;
        List<Pair> q = new ArrayList<>();
        q.add(new Pair(beginWord, 1, -1));
        visited.put(beginWord, 1);
        Pair cur ;
        int j = 0 ;

        int min_step = Integer.MAX_VALUE;
        boolean isFind = false;
        int pos;
        while( j < q.size()) {
            cur = q.get(j);
            if(cur.step > min_step) break;
            if(cur.first.equals(endWord)){
                if(!isFind) {
                    min_step = cur.step;
                }
                List<String> item = new ArrayList<>();
                item.add(cur.first);
                pos = cur.index;
                while(-1 != pos) {
                    item.add(q.get(pos).first);
                    pos = q.get(pos).index;
                }
                ret.add(item);
            }

            list = graph.get(cur.first);
            int size = 0;
            if(list != null ) size = list.size();
            for(int i=0; i<size; i++) {
                if(visited.containsKey(list.get(i)) == false || visited.get(list.get(i)) ==  cur.step + 1){
                    q.add(new Pair(list.get(i), cur.step + 1, j));
                    visited.put(list.get(i), cur.step + 1);
                }
            }
            j++;
        }
        return;
    }

    boolean connect(String word1, String word2){
        int count = 0;
        for(int i=0; i<word1.length(); i++) {
            if(word1.charAt(i) != word2.charAt(i)) count++;
        }
        return count==1;
    }

    void construct_graph(List<String> wordList, Map<String, List<String>> graph) {
        String istr;
        String jstr;
        for(int i=0; i<wordList.size()-1; i++) {
            for(int j=i+1; j<wordList.size(); j++){
                istr = wordList.get(i);
                jstr = wordList.get(j);
                if(connect(istr, jstr)){
                    graph.get(istr).add(new String(jstr));
                    graph.get(jstr).add(new String(istr));
                }
            }
        }

    }
};